intellij/src/saros/intellij/editor/annotations/SelectionAnnotation.java

```diff
@@ -109,6 +156,64 @@ void addLocalRepresentation(@NotNull Editor editor) {
     } else {
       log.warn("Could not create range highlighter for range " + annotationRange + " for " + this);
     }
+
+    addCaretRangeHighlighter(annotationRange, editor, user, file);
+  }
+
+  /**
+   * Adds a range highlighter for the caret annotation to the given editor and stores it in {@link
+   * #caretHighlighter}.
+   *
+   * <p>The highlighter is added at the start of the given annotation range if this is a backwards
+   * selection and at the end of the annotation range otherwise.
+   *
+   * @param annotationRange the annotation range of this annotations
+   * @param editor the editor of this annotation
+   * @param user the user of this annotation
+   * @param file the file of this annotation
+   */
+  private void addCaretRangeHighlighter(
+      @NotNull AnnotationRange annotationRange,
+      @NotNull Editor editor,
+      @NotNull User user,
+      @NotNull IFile file) {
+
+    int caretPosition;
+
+    if (isBackwardsSelection) {
+      caretPosition = annotationRange.getStart();
+    } else {
+      caretPosition = annotationRange.getEnd();
```

Comment:
I am not familiar with the annotation and range handling in IntelliJ.
How do you avoid that the caret is in the selected area?
* If I select the text "hello" from right to left (4-0) I would expect that `getStart` returns 0
and we have to create a caret left of 0 "|hello".
* If I select the text "hello" from left to right (0-4) I would expect that `getEnd` returns 4
and I have to create a caret right of 4 (or left of 5) "hello|".

However, with the current code, I would expect "hell|o" (or "h|ello") as a result. Do I have a wrong understanding of caret?

Reply:
Annotation ranges are always forward facing, meaning the positions returned by `getStart` and `getEnd` are the same for making selections "from left to right" or "from right to left".

In general, when thinking about text positions, you have to be careful not mixing up the text range and the "character array" view. For this text handling, we are talking about text ranges. For the purpose of selections, it is easiest to think about them as the positions between the characters. In you example, the range from 0-4 would only cover "hell" (0-0 covers "", 0-1 covers "h", 0-2 covers "he", and so on an so forth). As a result, when using the start (or end) of the range, we are always at the position before (or after) the first (or last) char of the range. As the caret annotation is displayed using a zero-width range highlighter, it will be displayed at that point before/after the first/last char. So 0-5 would be displayed as "hello|" or "|hello" respectively.

Reply:
Ah, thank you for the explanation :+1: